[fsdp] add an experimental allocator hook for buffers that participate in collective communication

jiayulu opened the pull request:
Summary: https://github.com/pytorch/pytorch/pull/147146

Test Plan: unit test

Differential Revision: D69694585




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_training.py:
```diff
@@ -1408,5 +1408,81 @@ def forward(self, imgs: torch.Tensor) -> torch.Tensor:
+        return min(torch.cuda.device_count(), 2)
+
+    @skip_if_lt_x_gpu(2)
+    def test_register_fsdp_forward_method(self):
```
nit: seems non-intuitive to use this test method to test the custom allocator

This test was testing a specific use case where the user does not invoke `module.forward()`.

weifengpy commented on torch/distributed/fsdp/_fully_shard/_fsdp_common.py:
```diff
@@ -179,3 +179,25 @@ def _cast_fp_tensor(dtype: torch.dtype, x: torch.Tensor) -> torch.Tensor:
+_fsdp_comm_allocator: Optional[_Allocator] = None
+
+
+def _set_fsdp_comm_allocator(allocator: _Allocator):
```
what happens we have have different mesh/pg for different `fully_shard`? maybe it's better to move `_set_fsdp_comm_allocator` and `_get_fsdp_comm_allocator` to FSDPModule https://github.com/pytorch/pytorch/blob/81719ebde3f8ac0953a69b919809b39c2918a0a7/torch/distributed/fsdp/_fully_shard/_fully_shard.py#L516

example usage of _set_fsdp_comm_allocator
```
for layer in model.layers:
    fully_shard(layer, mesh=)
    layer._set_fsdp_comm_allocator(layer._get_fsdp_state()._fsdp_param_group.shard_process_group)
fully_shard(model, mesh=)
model._set_fsdp_comm_allocator(model._get_fsdp_state()._fsdp_param_group.shard_process_group)
```

jiayulu replied:
@weifengpy once the user call `_set_fsdp_comm_allocator(custom_allocator)`, `custome_allocator` can be used by all nccl PGs, i.e. it's not tied to a PG.

weifengpy replied:
for 2d mesh, do we need to split `_set_fsdp_comm_allocator` into 
* `_set_ag_rs_allocator` for all-gather and reduce-scatter
* `_set_ar_allocator` for all-reduce

weifengpy replied:
> once the user calls _set_fsdp_comm_allocator(custom_allocator), custome_allocator can be used by all nccl PGs, i.e. it's not tied to a PG

ah, that's good to know

jiayulu replied:
> for 2d mesh, do we need to split `_set_fsdp_comm_allocator` into
> 
> * `_set_ag_rs_allocator` for all-gather and reduce-scatter
> * `_set_ar_allocator` for all-reduce

@weifengpy no we don't need to split it into two calls.

jiayulu commented:
/easycla

weifengpy reviewed (approved):
fix the linter? it looks good

weifengpy commented:
```
File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_common.py", line 188
    ...
IndentationError: unexpected indent
```

the CI error is related for sure this time. maybe run unit test to repro?

jiayulu committed:
[fsdp] add an experimental allocator hook for buffers that participate in collective communication (#149150)

Summary:

https://github.com/pytorch/pytorch/pull/147146

Test Plan: unit test

Differential Revision: D69694585

weifengpy reviewed (changes requested):
there are choices of UX
* option 1: fully_shard(mempool=), favored by Luca
* option 2: _set_fsdp_comm_allocator, _get_fsdp_comm_allocator fully_shard(allocator=), favored by this PR
* option 3: ProcessGroupNCCL::allocateTensor(...) without exposing it to fully_shard, favored by Ke

need to align on this. Personally I can green light the PR with _set_fsdp_comm_allocator, _get_fsdp_comm_allocator. we can consolidate to option 3 later
